libsolidity/analysis/SyntaxChecker.cpp

```diff
@@ -183,6 +186,58 @@ bool SyntaxChecker::visit(Throw const& _throwStatement)
 	return true;
 }
 
+bool SyntaxChecker::visit(Literal const& _literal)
+{
+	// Generic checks no matter what base this number literal is of:
+	if (_literal.value().back() == '_')
+	{
+		m_errorReporter.syntaxError(_literal.location(), "Invalid use of underscores in number literal. No trailing underscores allowed.");
+		return true;
+	}
+
+	if (_literal.value().find("__") != ASTString::npos)
+	{
+		m_errorReporter.syntaxError(_literal.location(), "Invalid use of underscores in number literal. Only one consecutive underscores between digits allowed.");
+		return true;
+	}
+
+	if (!_literal.isHexNumber())
+		// No special validation on grouped decimals yet.
+		return true;
+
+	// We have a hex literal. Do underscore validation
+	solAssert(_literal.value().substr(0, 2) == "0x", "");
+
+	ASTString value = _literal.value().substr(2); // Skip the 0x
+	vector<ASTString> parts;
+	boost::split(parts, value, boost::is_any_of("_"));
+
+	if (parts.size() == 1) // no underscores
+		return true;
+
+	if (parts.front().size() > 4)
+		m_errorReporter.syntaxError(_literal.location(), "Invalid use of underscores in hex literal. The first part " + parts[0] + " is too large. Group up by thousands.");
+
+	// Everything except first and last part must be 4 chars in length
+	for (size_t i = 1; i + 1 < parts.size(); ++i)
```

Comment:
Maybe ``i < parts.size() - 1`` would be easier to read.

Reply:
I agree.